### PR TITLE
refactor(api): VALIDATE macros

### PR DESCRIPTION
**Pull request by justinmk**
# Problem

- API validation involves too much boilerplate.
- API validation errors are not consistently worded.

# Solution

Introduce some macros. Currently these are clumsy, but they at least help with consistency and avoid some nesting.

### CODE REVIEW

File: src/nvim/api/vim.c [hunk 1]
```diff
@@ -1964,9 +1952,7 @@ Array nvim__inspect_cell(Integer grid, Integer row, Integer col, Arena *arena, E
     if (wp != NULL && wp->w_grid_alloc.chars != NULL) {
       g = &wp->w_grid_alloc;
     } else {
-      api_set_error(err, kErrorTypeValidation,
-                    "No grid with the given handle");
-      return ret;
+      VALIDATE_INT(true, "grid handle", grid, return ret);
```
Comment (lewis6991) [inline src/nvim/api/vim.c:1955]: This style will hurt treesitter. Best try and phrase the macros like:

```c
VALIDATE_INT(true, "grid handle", grid) {
   return ret;
};
```

Comment (justinmk) [reply src/nvim/api/vim.c:1955]: yeah that avoids the non-obvious control flow too.

File: src/nvim/api/private/validate.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#ifndef NVIM_API_PRIVATE_VALIDATE_H
```
Comment (justinmk) [inline src/nvim/api/private/validate.h:1]: introducing a validate.h module in anticipation that we will have more sophisticated validation things for C API in the future.

File: src/nvim/api/vim.c [hunk 3]
```diff
@@ -2009,20 +1980,16 @@ Boolean nvim_del_mark(String name, Error *err)
...
-  } else {
-    api_set_error(err, kErrorTypeValidation,
-                  "Only file/uppercase marks allowed, invalid mark name: '%c'",
-                  *name.data);
-  }
+  VALIDATE_S((ASCII_ISUPPER(*name.data) || ascii_isdigit(*name.data)),
+             "mark name (must be file/uppercase)", name.data, {
+    return res;
+  });
+  res = set_mark(NULL, name, 0, 0, err);
```
Comment (justinmk) [inline src/nvim/api/vim.c:1992]: this is a nice win from inverting the validation condition

File: src/nvim/api/private/validate.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#define NVIM_API_PRIVATE_VALIDATE_H
+
+#include "nvim/api/private/defs.h"
+#include "nvim/api/private/helpers.h"
+
+#define VALIDATE_INT(cond, name, n_, code) \
+  do { \
```
Comment (lewis6991) [inline src/nvim/api/private/validate.h:8]: Are the do-while wrappers needed since it is only an if-block?

File: src/nvim/api/private/validate.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      if (strequal(str_, "")) { \
+        api_set_error(err, kErrorTypeValidation, "Invalid " name); \
+      } else { \
+        api_set_error(err, kErrorTypeValidation, "Invalid " name ": '%s'", str_); \
+      } \
+      code; \
+    } \
+  } while (0)
+
+#define VALIDATE_T(name, expected_t, actual_t, code) \
```
Comment (lewis6991) [inline src/nvim/api/private/validate.h:27]: Is `VALIDATE_EQ` a better name?

File: src/nvim/api/private/validate.h (hunk 4 above)
Comment (lewis6991) [reply src/nvim/api/private/validate.h:8]: Ah, I see why it is needed.

File: src/nvim/api/private/validate.h (hunk 5 above)
Comment (lewis6991) [reply src/nvim/api/private/validate.h:27]: Ah wait no it isn't.

**Pull request merged**